daemon/cluster/controllers/plugin/controller.go

```diff
@@ -1,32 +1,140 @@
+// Plugins are managed as a singleton object with a desired state (different from containers).
+// With the the plugin controller instead of having a strict create->start->stop->remove
+// task lifecycle like containers, we manage the desired state of the plugin and let
+// the plugin manager do what it already does and monitor the plugin.
+// We'll also end up with many tasks all pointing to the same plugin ID.
+type Controller struct {
+	backend   Backend
+	spec      runtime.GenericPluginPayload
+	id        string
+	serviceID string
+
+	// hook used to signal tests that we're actually watching
+	signalWaitReady func()
+}
+
+// Backend is the interface for interacting with the plugin manager
+// Controller actions are passed to the configured backend to do the real work.
+type Backend interface {
+	Disable(name string, config *enginetypes.PluginDisableConfig) error
+	Enable(name string, config *enginetypes.PluginEnableConfig) error
+	Remove(name string, config *enginetypes.PluginRmConfig) error
+	Pull(ctx context.Context, ref reference.Named, name string, metaHeaders http.Header, authConfig *enginetypes.AuthConfig, privileges enginetypes.PluginPrivileges, outStream io.Writer, opts ...plugin.CreateOpt) error
+	Upgrade(ctx context.Context, ref reference.Named, name string, metaHeaders http.Header, authConfig *enginetypes.AuthConfig, privileges enginetypes.PluginPrivileges, outStream io.Writer) error
+	Get(name string) (*v2.Plugin, error)
+	SubscribeEvents(buffer int, events ...plugin.Event) (eventCh <-chan interface{}, cancel func())
+}
 
 // NewController returns a new cluster plugin controller
-func NewController() (*Controller, error) {
-	return &Controller{}, nil
+func NewController(backend Backend, t *api.Task) (*Controller, error) {
+	spec, err := readPayload(t)
+	if err != nil {
+		return nil, err
+	}
+	return &Controller{backend: backend, spec: spec, serviceID: t.ServiceID}, nil
 }
 
 // Update is the update phase from swarmkit
 func (p *Controller) Update(ctx context.Context, t *api.Task) error {
 	logrus.WithFields(logrus.Fields{
 		"controller": "plugin",
 	}).Debug("Update")
+
 	return nil
 }
 
+func readPayload(t *api.Task) (runtime.GenericPluginPayload, error) {
+	var cfg runtime.GenericPluginPayload
+
+	generic := t.Spec.GetGeneric()
+	if err := proto.Unmarshal(generic.Payload.Value, &cfg); err != nil {
+		return cfg, errors.Wrap(err, "error reading plugin config")
+	}
+	return cfg, nil
+}
+
 // Prepare is the prepare phase from swarmkit
-func (p *Controller) Prepare(ctx context.Context) error {
+func (p *Controller) Prepare(ctx context.Context) (err error) {
 	logrus.WithFields(logrus.Fields{
 		"controller": "plugin",
 	}).Debug("Prepare")
+
+	// TODO(@cpuguy83): support digest pinning
+	remote, err := reference.ParseNormalizedNamed(p.spec.Spec.Remote)
+	if err != nil {
+		return errors.Wrapf(err, "error parsing remote reference %q", p.spec.Spec.Remote)
+	}
+
+	if p.spec.Spec.Name == "" {
```

Comment:
I don't see us ever checking that `p.spec.Spec` is non-nil. I'd recommend adding `[(gogoproto.nullable) = false]` to the line where the `spec` field is declared in the .proto file. This will cause the field to be declared as a `PluginSpec`, not `*PluginSpec`. If the field is missing from the serialized proto for some reason, the values in the `PluginSpec` will be empty, but the code trying to access it won't crash. We do this in swarmkit for all specs, because they are never expected to be missing, so avoiding the pointer indirection avoids the need for a lot of nil checks that would otherwise be needed to avoid a crash when dealing with improper or corrupted input.